### ISSUE TITLE
[feature request] Circular Convolution Function (Convolution with circular padding) 

### ISSUE THREAD

**Issue by YuesongNan** [pytorch/pytorch#6124]
In the realm of image processing, Circular Convolution is common used because it is suitable to do FFT. 
Circular Convolution means that firstly padding the tensor with circular boundary and then do the convolution. It works like _scipy.ndimage.filters.convolve(x,ker,mode='wrap')_ in Scipy or _imfilter(x,ker,'circular','conv')_ in Matlab. 
Up to now, Pytorch does not support this operation. In _torch.nn.functional.con3d_, only zero padding are supported.
An alternative solution is padding the tensor first. However, in _torch.nn.functional.pad_, only "constant", "replicate" and "replicate" modes are supported. 
Please do consider this request as the circular convolution is a very basic operation in image processing and a fast implementation is needed. 
Thank you!

**Comment by zou3519** [comment]
What do you mean by `circular boundary`?

**Comment by fmassa** [comment]
@zou3519 circular boundary means that you pad on the right with the pixels from the left of the image, as if you had concatenated the image on the right. The same applies for the bottom of the image.

**Comment by zou3519** [comment]
Ah gotcha, thanks for the clarification @fmassa.  @YuesongNan as a workaround you can try what I suggested here: https://github.com/pytorch/pytorch/issues/3858 to pad your tensor before a convolution.

**Comment by fmassa** [comment]
Another possible implementation would be to modify `im2col` to handle this case.
When out of boundary, instead of setting to 0 in [here](
https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/Col2Im.c#L77-L78)
you'd instead return a the value of the input at the position.

**Comment by YuesongNan** [comment]
@zou3519 Thank you for your reply. I am actually trying this method. But I don't know whether using `torch.cat` for padding will have the same speed performance as modification on the source code of the convolution. 
@fmassa Thank you for help. I am trying to understand this code and do corresponding modification.

**Comment by SsnL** [comment]
Thanks for your request. This is quite reasonable, but there will be some issues to work around with cudnn. We will put this on our list of task to do. :)

**Comment by ngimel** [comment]
Cudnn will do just fine if there is a preceding explicit padding layer like @zou3519 suggested, but making it actually truly support circular convolution is a different discussion.

**Comment by YuesongNan** [comment]
@SsnL Thank you for your attention and I am glad to hear that this feature will be incorporated in the future. By the way, I have some small suggestions for convolution implementation. 
1. in `torch.nn.functional` there is the implementation of `conv2d`which actually a correlation. It is fine to use correlation instead of convolution in deep learning setting, but for scientific computing, it makes confusion for new comers. I was thinking if it was possible for implement another true convolution with just flip the kernel. 
2. Still in scientific computing setting, I was thinking if was possible to provide convolution with cuda implementation for util function as well. This implementation will avoid transforming 2d `torch.Tensor` type to `Variable` type with 4d structure ([input, output, image_size, image_size]). 
Thank you again for your help. Above suggestions are just for you as reference.

**Comment by soumith** [comment]
@YuesongNan 

about (1), you are right that conv2d is correlation, but this is terminology that is (unfortunately) adopted by deep learning community. We will likely keep this terminology (for better or worse)

**Comment by Naruto-Sasuke** [comment]
Hope it will be added ASAP.

**Comment by thecortex** [comment]
I have also wanted to test it: https://arxiv.org/pdf/1706.05534.pdf

**Comment by ezyang** [comment]
@isseu has kindly agreed to take this on. Here are some notes on how to implement:

1. We have a simple implementation of convolution by using im2col/vol2col to convert matrices into a Toeplitz matrix, and then doing a matrix multiplication to carry out the convolution. (If you don't see how this works, spend a little time with the matrices until you convince yourself it works :)
2. The CPU implementation of im2col lives at `aten/src/THNN/generic/Col2Im.c`. Adjust it to optionally generate a *circular* Toeplitz matrix. For testing, you may find it easier to first make the function unconditionally generate a circular Toeplitz matrix, and ensure that all convolutions are circular, before carrying on. (Be aware that you need to make sure the THNN kernel is actually being called; there is logic to select which convolution kernel we use, and it does not always select THNN.)
3. Thread your change to the call sites of im2col, which actually perform the convolution. After you fix the function sites, you'll need to modify `aten/src/ATen/nn.yaml` to adjust our code generation to handle the new signature. You might find it easiest to create a new function which performs circular convolution, and call into some generic function that takes the circular convolution as a flag.
4. Adjust `_convolution` (defined in `aten/src/ATen/native/native_functions.yaml`) to take a flag trigger circular padding, and adjust the kernel selection logic to prefer THNN implementation when circular padding is requested.
5. Adjust the backwards formula for convolution, so that it works in the circular padding case. (This should not be too difficult.)
5. Write tests for it
6. Do this again for the CUDA implementation

**Comment by gchanan** [comment]
Just a note on 5.  You can find how the backward formula definitions are linked up to the forward in `derivatives.yaml`.

**Comment by harpone** [comment]
In case anyone is looking for a faster circular padding, I wrote a small function here: https://github.com/pytorch/pytorch/issues/3858#issuecomment-432801360

**Comment by gycn** [comment]
Has there been any progress on this issue?

**Comment by nikita-astronaut** [comment]
Buys, the periodic padding is VERY needed!

**Comment by fmassa** [comment]
Can't we just use circular padding (from `torch.nn.functional.pad` and apply a convolution without padding afterwards?

**Comment by jzazo** [comment]
Is this matter resolved and published by this [pull request](https://github.com/pytorch/pytorch/pull/17240)?
Is this other [request](https://github.com/pytorch/pytorch/issues/3858) also solved by the same pull?

Are there unaddressed matters for circular convolution? Thanks.

**Comment by Chillee** [comment]
Pretty sure these are both resolved @jzazo . I'll close these issues. Reopen if there's something I missed.
